zebra-network/src/meta_addr.rs

```diff
@@ -160,135 +823,244 @@ impl MetaAddr {
     ///
     /// Otherwise:
-    /// - malicious peers could interfere with other peers' `AddressBook` state,
+    /// - malicious peers could interfere with other peers' [`AddressBook`] state,
     ///   or
     /// - Zebra could advertise unreachable addresses to its own peers.
-    pub fn new_responded(addr: &SocketAddr, services: &PeerServices) -> MetaAddr {
-        MetaAddr {
-            addr: *addr,
-            services: *services,
-            last_seen: Utc::now(),
-            last_connection_state: Responded,
-        }
+    pub fn update_responded(addr: SocketAddr, services: PeerServices) -> MetaAddrChange {
+        UpdateResponded { addr, services }
     }
 
-    /// Create a new `MetaAddr` for a peer that we want to reconnect to.
-    pub fn new_reconnect(addr: &SocketAddr, services: &PeerServices) -> MetaAddr {
-        MetaAddr {
-            addr: *addr,
-            services: *services,
-            last_seen: Utc::now(),
-            last_connection_state: AttemptPending,
-        }
+    /// Update an [`AddressBook`] entry when a peer connection fails.
+    pub fn update_failed(addr: SocketAddr, services: Option<PeerServices>) -> MetaAddrChange {
+        UpdateFailed { addr, services }
     }
 
-    /// Create a new `MetaAddr` for a peer's alternate address, received via a
-    /// `Version` message.
-    pub fn new_alternate(addr: &SocketAddr, services: &PeerServices) -> MetaAddr {
-        MetaAddr {
-            addr: *addr,
-            services: *services,
-            last_seen: Utc::now(),
-            last_connection_state: NeverAttemptedAlternate,
-        }
+    /// Update an [`AddressBook`] entry when a peer connection shuts down.
+    pub fn update_shutdown(addr: SocketAddr, services: Option<PeerServices>) -> MetaAddrChange {
+        UpdateShutdown { addr, services }
     }
 
-    /// Create a new `MetaAddr` for our own listener address.
-    pub fn new_local_listener(addr: &SocketAddr) -> MetaAddr {
-        MetaAddr {
-            addr: *addr,
+    /// Add or update our local listener address in an [`AddressBook`].
+    ///
+    /// See [`AddressBook::get_local_listener`] for details.
+    pub fn new_local_listener(addr: SocketAddr) -> MetaAddrChange {
+        NewAlternate {
+            addr,
+            // Note: in this unique case, the services are actually trusted
+            //
             // TODO: create a "local services" constant
-            services: PeerServices::NODE_NETWORK,
-            last_seen: Utc::now(),
-            last_connection_state: Responded,
+            untrusted_services: PeerServices::NODE_NETWORK,
         }
     }
 
-    /// Create a new `MetaAddr` for a peer that has just had an error.
-    pub fn new_errored(addr: &SocketAddr, services: &PeerServices) -> MetaAddr {
-        MetaAddr {
-            addr: *addr,
-            services: *services,
-            last_seen: Utc::now(),
-            last_connection_state: Failed,
-        }
```

Comment:
Security: Zebra should stop gossiping failure times as last_seen times #1868

Part 1c: Stop using failure times as last seen times